### PR TITLE
Merge readable and writable buffer component iteration

### LINKED ISSUE #12609: Hard to get direct memory address for netty5 Buffer
I want to get a direct memory address for Buffer where `writerOffset() == readerOffset()` but current design of Buffer in netty5 is hard.
If `writerOffset() == readerOffset() && writerOffset() == capacity()` then `forEachReadable` and `forEachWritable` do not `foreach`.
If there any readable or writable bytes then this _solution_ add something similar of netty4 ByteBuf memoryAddress:
```
var address = 0L
buf.forEachReadable(
  0,
  (_, component) =>
    address = component.readableNativeAddress() - buf.readerOffset()
    false
)
if address == 0 then
  buf.forEachWritable(
    0,
    (_, component) =>
      address = component.writableNativeAddress() - buf.writerOffset()
      false
  )
address match
  case 0L => throw BufferMemoryAddressIsZero
  case _ => address
```

Why this is so hard in netty5? Previous ByteBuf'er is easy to use with native code (JEP-419).

### Expected behavior

When I parsing a Buffer and get to the end I want to get raw memoryAddress of my single offheap Buffer and pass it to native code with my offset and length variables by myself, for example to check integrity in STUN protocol.

### Actual behavior

forEachReadable and forEachWritable do not foreach without changing writerOffset or readerOffset. And there no ability to get direct memory address without a proxy class like SingleComponentIterator.

### Netty version

netty-5.0.0.Alpha3

### PR DESCRIPTION
**Pull request by chrisvest**
Motivation:
It should be possible to iterate the components of a buffer, even when it has no readable or writable bytes.
For the purpose of pointer arithmetic, it should also be possible to get a base pointer for off-heap buffers.
It additionally turns out that unifying ReadableComponent and WritableComponent into a single BufferComponent, allow us to make the iteration entirely allocation free for non-composite buffers.
This in turn allow us to simplify the API by removing the internal iteration methods and only keep a single forEachComponent iteration method.

Modification:
- Remove WritableComponent and ReadableComponent, and their processors.
- And in their place have a single BufferComponent which contain both of the APIs from the readable and writable components.
- Add a baseNativeAddress method to BufferComponent.
- Change all usages sites of both the internal and external versions of forEachReadable/Writable, to use the new forEachComponent.
- Add firstReadable and firstWritable convenience methods to ComponentIterator, which will skip over components that are not readable or writable, respectively.
- Add nextReadable and nextWritable methods to ComponentIterator.Next, which do the same.
- Change places like EngineWrapper, IovArray, etc. to no longer rely on internal iteration but use external iteration instead.
- Remove processor classes that only existed to be used for internal iteration.
- Read-only buffers no longer have any "writable bytes".

Result:
Iterating the components of non-composite buffers is now always allocation-free; no capturing-lambda or iterator objects are constructed.
The APIs are also simpler; fewer methods and fewer classes.

Fixes #12609 and closes #12611

### CODE REVIEW

**Comment by chrisvest** [comment]
@timothyklim FYI

File: buffer/src/main/java/io/netty5/buffer/api/Buffer.java [hunk 1]
```diff
@@ -1051,9 +1054,9 @@ default Buffer split() {
 
     /**
      * Get the number of "components" in this buffer, that are writable. These are the components that would be
-     * processed by {@link #forEachWritable(int, WritableComponentProcessor)}. For composite buffers, this is the
-     * number of transitive constituent buffers that are writable, while non-composite buffers only have at most one
-     * writable component.
+     * processed by {@link #forEachComponent()}, by followin the {@link ComponentIterator#firstWritable()} and
```
Comment (hyperxpro) [inline buffer/src/main/java/io/netty5/buffer/api/Buffer.java:1057]:
```suggestion
     * processed by {@link #forEachComponent()}, by following the {@link ComponentIterator#firstWritable()} and
```

**Review by hyperxpro — CHANGES_REQUESTED** [review]

File: buffer/src/main/java/io/netty5/buffer/api/Buffer.java [hunk 2]
```diff
@@ -1142,95 +1102,11 @@ default Buffer split() {
...
-     * this buffer instance.
-     *
-     * @return A component iterator of {@linkplain ReadableComponent readable components}.
-     * @param <T> An intersection type that presents both the {@link ReadableComponent} interface,
+     * @return A component iterator of {@linkplain BufferComponent readable components}.
+     * @param <T> An intersection type that presents both the {@link BufferComponent} interface,
      *          <em>and</em> the ability to progress the iteration via the {@link Next#next()} method.
      */
-    <T extends WritableComponent & ComponentIterator.Next> ComponentIterator<T> forEachWritable();
+    <T extends BufferComponent & ComponentIterator.Next> ComponentIterator<T> forEachComponent();
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty5/buffer/api/Buffer.java:1109]: I wonder if we should change this to something like:

```
R <R> forEachComponent(Function<R, T extends BufferComponent & ComponentIterator.Next> ComponentIterator<T>>)
```

This way we could make it safer to ensure that the native address "never escapes" while iteration is in progress.

**Review by normanmaurer — APPROVED** [review]
I like this change a lot! Just one question which we can also address as a followup.

File: buffer/src/main/java/io/netty5/buffer/api/Buffer.java (hunk 2 above)
Comment (normanmaurer) [reply buffer/src/main/java/io/netty5/buffer/api/Buffer.java:1109]: We might also want to consider adding an extra method that takes a `ToLongFunction ` to reduce the overhead of boxing/unboxing.

Comment (chrisvest) [reply buffer/src/main/java/io/netty5/buffer/api/Buffer.java:1109]: This sounds like internal iteration again, where we often end up allocating objects for capturing lambdas.
And the ability to return data creates a path for the native address to escape, so it doesn't sound much safer than the try-with-resources we currently have.

**Comment by chrisvest** [comment]
Looking into the epoll test failure that showed up after the rebase.

**Pull request merged**